eth/stagedsync/stage_witness.go

```diff
@@ -0,0 +1,424 @@
+			}
+		}
+
+		// Write the witness buffer against the corresponding block number
+		logger.Debug("Writing witness to db", "block", blockNr)
+		err = WriteChunks(rootTx, kv.Witnesses, Uint64ToBytes(blockNr), buf.Bytes())
+		if err != nil {
+			return fmt.Errorf("error writing witness for block %d: %v", blockNr, err)
+		}
+
+		// Update the stage with the latest block number
+		s.Update(rootTx, blockNr)
+
+		// If we're overlimit, delete oldest witness
+		oldestWitnessBlock, _ := FindOldestWitness(rootTx, kv.Witnesses)
+		if blockNr-oldestWitnessBlock+1 > cfg.maxWitnessLimit {
+			// If the user reduces `witness.limit`, we'll need to delete witnesses more than just oldest
+			deleteFrom := oldestWitnessBlock
+			deleteTo := blockNr - cfg.maxWitnessLimit
+			logger.Debug("Reached max witness limit, deleting oldest witness", "from", deleteFrom, "to", deleteTo)
+			for i := deleteFrom; i <= deleteTo; i++ {
+				err = DeleteChunks(rootTx, kv.Witnesses, Uint64ToBytes(i))
+				if err != nil {
+					return fmt.Errorf("error deleting witness for block %d: %v", i, err)
+				}
+			}
+		}
+
+		logger.Info(fmt.Sprintf("[%s] Generated witness", logPrefix), "block", blockNr, "len", len(buf.Bytes()))
+	}
+
+	logger.Info(fmt.Sprintf("[%s] Done Witness Generation", logPrefix), "until", to)
+
+	return nil
+}
+
+func RewindStagesForWitness(batch *membatchwithdb.MemoryMutation, blockNr uint64, cfg *WitnessCfg, regenerateHash bool, ctx context.Context, logger log.Logger) (*membatchwithdb.MemoryMutation, *trie.RetainList, error) {
+	rl := trie.NewRetainList(0)
+
+	// Rewind the 'HashState' and 'IntermediateHashes' stages to previous block
+	unwindState := &UnwindState{ID: stages.HashState, UnwindPoint: blockNr - 1}
+	stageState := &StageState{ID: stages.HashState, BlockNumber: blockNr}
+
+	hashStageCfg := StageHashStateCfg(nil, cfg.dirs, false)
+	if err := UnwindHashStateStage(unwindState, stageState, batch, hashStageCfg, ctx, logger); err != nil {
+		return nil, nil, err
+	}
+
+	unwindState = &UnwindState{ID: stages.IntermediateHashes, UnwindPoint: blockNr - 1}
+	stageState = &StageState{ID: stages.IntermediateHashes, BlockNumber: blockNr}
+
+	if !regenerateHash {
+		interHashStageCfg := StageTrieCfg(nil, false, false, false, "", cfg.blockReader, nil, false, nil)
+		err := UnwindIntermediateHashes("eth_getWitness", rl, unwindState, stageState, batch, interHashStageCfg, ctx.Done(), logger)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else {
+		_ = batch.ClearBucket(kv.TrieOfAccounts)
+		_ = batch.ClearBucket(kv.TrieOfStorage)
+	}
+
+	return batch, rl, nil
+}
+
+func GenerateWitness(tx kv.Tx, block *types.Block, prevHeader *types.Header, tds *state.TrieDbState, trieStateWriter *state.TrieStateWriter, statedb *state.IntraBlockState, getHashFn func(n uint64) libcommon.Hash, cfg *WitnessCfg, regenerateHash bool, ctx context.Context, logger log.Logger) (*trie.Witness, error) {
+	blockNr := block.NumberU64()
```

Comment:
We need to make a call to `statedb.SetDisableBalanceInc(true)` in the beginning of the function. Otherwise, some trie nodes will be missed in the witness. See this commit: https://github.com/0xPolygonZero/erigon/commit/f0efcac0735a081dceac59ec297370b5ef3bce72.

Reply:
I see, thanks for pointing it out. Will handle it.